ci/main.py

```diff
@@ -3,61 +3,119 @@
+    Parameters
+    ----------
+    p : str
+        directory path to the recipes
+    channel : str
+        the channel where the packages will be uploaded.
+    '''
     for root, dirs, files in os.walk(p):
         has_recipe = 'meta.yaml' in files
         if not dirs and has_recipe:
             with open(os.path.join(root, 'meta.yaml')) as f:
-                pkg = yaml.load(f)['package']
+                print('PROCESSING {0}'.format(root))
+                meta = yaml.load(f)
+                try:
+                    build_n = meta['build']['number']
+                except KeyError:
+                    # if not specified, the build number is 0 by default
+                    build_n = 0
+                pkg = meta['package']
                 name = pkg['name']
                 version = pkg['version']
-                build(name, version, root)
-                if os.environ['TRAVIS_SECURE_ENV_VARS'] == 'true':
-                    upload(name, version)
-                else:
-                    print("Uploading not available in Pull Requests")
+                if check(name, version, build_n, channel):
+                    build(name, version, root)
+                    if os.environ['TRAVIS_SECURE_ENV_VARS'] == 'true':
+                        upload(name, version, channel)
+                    else:
+                        print("Uploading not available in Pull Requests")
 
 
-def build(name, version, root):
-    print("Building package: {0}-{1}".format(name, version))
-    # Quote is need in case the root path has spaces in it.
-    build_cmd = 'conda build "%s"' % root
-    print('BUILDING COMMAND: {0}'.format(build_cmd))
-    check_call(build_cmd, shell=True)
-
-
-def upload(name, version):
-    check_cmd = ('conda search --json --override-channels '
-                 '-c {0} --spec {1}={2}').format(
-                     CHANNEL, name, version)
-    print('CHECKING COMMAND: {0}'.format(check_cmd))
-    try:
-        out = check_output(check_cmd, shell=True)
-    except CalledProcessError as e:
-        # This built version does not exist in the channel
-        out = e.output
+def check(name, version, build_n, channel):
+    '''Check if we want to build & upload a package.
+
+    It checks package name, package version, and package build number
+    sequentially to decide whether to return True or not.
+
+    Parameters
+    ----------
+    name : str
+        package name.
+    version : str
+        package version.
+    build_n : int
+        build number
+    channel : str
+        the channel to check the previous built.
+
+    Returns
+    -------
+    boolean
+        If a previous package in the channel has the same name, the same
+        version, and an equal or higher build number, then return False;
+        otherwise, return True.
+    '''
+    cmd = ('conda search --json --override-channels '
+           '-c {0} --spec {1}').format(channel, name)
+    print('CHECK COMMAND: {0}'.format(cmd))
```

Comment:
These `print` statements almost look like they should be log statements via `logging`?